`-Xlint` now includes `-Xlint:named-booleans` to lint unnamed boolean literal args

som-snytt opened the pull request:
## Summary

Since boolean literals are not expressive of intention when passed as arguments, suggest using a named argument. Instead of `throwBalloons(true, false)`, the intent is clearer with:

```
throwBalloons(waterFilled = true, now = false)
```
The lint should warn only for boolean flags which may invert the meaning of the method call. It only warns if the parameter is of type `Boolean`, so that `Some(true)` or `getOrElse(false)` does not warn.

## Enabling/disabling

The new lint is part of the package of default lints enabled when you pass `-Xlint`.

To enable the new lint in isolation, use `-Xlint:booleanLiterals`.

To disable the new lint while leaving other default lints enabled, use `-Xlint:-booleanLiterals`.

## Details

As a special exemption to reduce false positives, the lint will not warn if there is only one boolean parameter in first position, such as `assert(false, msg)`. Consequently, methods taking a single boolean parameter are expected to have a meaningful name that makes the semantics of the parameter obvious, such as `Option.when(true)`.

Compare:
```
threatAssessment(launchMissiles = true)  // the parameter name is not enforced
launchMissilesWhen(true) // improved

invoke(arg: A, async = true) // annoying name? try a default arg, overload, or methods with explicit names
invoke(arg) // sync
invokeAsync(arg) // obviously

val expected = Array(true, false, true) // ok as the second parameter is repeating
```

Since class names tend to be nouns instead of specific verbs, constructors (and also proxies, i.e., case class `apply`) will warn:
```
SortedList(reverse = true)
```
It's terrible to rely on naming for expressiveness or (the illusion of) type safety, as opposed to actual types that ensure you can't accidentally swap `start` and `end` values (which one assumes is unlikely only if one reads from left to right or if one does not routinely mix things up), but this lint provides some "support" for the common case. By "support", we understand that it "enables" this bad behavior.

The lint will not warn if the parameter is annotated `deprecatedName`:
```
def addIf(@deprecatedName b: Boolean, v: Value, deduplicate: Boolean): this.type // naming `b` is extraneous
```

Note that it does not warn for tuples, but in future, named tuples should be encouraged for this idiom:
```
def f = if (cond) (x, true) else (y, false)
```

Possible future improvements could be to warn for other anonymous literals such as `null`, primitives such as `42`, or special cases of zero or empty such as `None`, `Nil`, or `Set.empty`.

## Further details

The corresponding parameter must be of `Boolean` type. `Some(true)` does not warn; nor does `bs.fold(false)`.

Exclude methods with exactly one boolean parameter which is in first position, like `assert`. Such methods are expected to be tastefully named, so that the boolean needn't be.

Parameters marked `@deprecatedName` are excluded.

Stratagems include a judicious mix of use named arg, sprinkle deprecatedName, or leverage default args (by omitting the literal arg in favor of a default). Also nowarn. These workarounds are mostly applied in test code.

som-snytt committed:
Lint unnamed boolean literal args

som-snytt commented on src/compiler/scala/tools/reflect/ToolBoxFactory.scala:
```diff
@@ -184,10 +184,10 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
             currentTyper.silent(_.typed(expr, mode, pt), reportAmbiguousErrors = false) match {
               case analyzer.SilentResultValue(result) =>
-                trace("success: ")(showAttributed(result, true, true, settings.Yshowsymkinds.value))
+                trace("success: ")(showAttributed(result, printKinds = settings.Yshowsymkinds.value))
```
this was a bug. it suggests that requiring names can be broadly useful. (Here, the many booleans are not distinguished by type.)

som-snytt commented:
Wasn't there a merged retronym PR https://github.com/scala/scala/pull/2164

I agree with the old paulp comment that it's better to improve API, for example `thingWithoutExtra` instead of `thing(extra = false)`.

Also there is no lint to make sure I did not accidentally incur an arg block, by writing named args out of order.

There is a ticket about the rewrite not using temporaries of constants, so an application that only switches constants could skip the block and maybe warn. `f(p(), y = true, x = false)`

The possibly legit use of long tuples of booleans is in tests, but it would be nice to have a cleaner way to create a table and use it, for that use case.

som-snytt commented:
The festival held once every ten years is called The Gathering of the Names, in memory of all those unnamed program elements which shall go unnamed here.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1714,6 +1714,33 @@ abstract class RefChecks extends Transform {
+        }
+        if (settings.lintNamedBooleans && lintedBoolean)
+          foreach2(args, fn.symbol.paramss.head)((arg, param) => arg match {
+            case t @ Literal(Constant(_: Boolean)) if t.hasAttachment[UnnamedArg.type] && !innommable(param) && !isRepeated(param) =>
```
Did you consider `&& param.tpe.typeSymbol == BooleanClass`? This would cover the special cases for Option / Array / etc above. Would it eliminate some useful warnings?

For example this PR adds `newLiteral(value = true)` above, which I think is not necessary.

som-snytt replied:
I added deprecatedName to a couple of methods such as `newLiteral` and `tpd.Constant`, which take a `value` which is just the value. Also `boolconst`.

lrytz replied:
It still seems to me that warning only if the paramter type is a boolean would be right. This will also come up in user code.

som-snytt replied:
Now that the dust has settled, I see your point. I was blinded by my decision about useful names, but only boolean flags have useful names anyway. I will follow up forthwith.

Hypothetical escape hatch `def f[A <: Boolean](b: A)`. IIRC `BooleanSetting` takes a primitive that is boxed anyway.

Also solves `fold`.

lrytz commented on src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:
```diff
@@ -79,10 +79,10 @@ trait ContextOps { self: TastyUniverse =>
   }
 
-  @inline final def assert(assertion: Boolean, msg: => Any): Unit =
+  @inline final def assert(@deprecatedName assertion: Boolean, msg: => Any): Unit =
```
Ah, a secret handshake!

lrytz reviewed (commented):
This looks great, thanks!

som-snytt commented:
I considered whether other literals suffer the same issue (lack of expressiveness): `42` (which may be linted as a "magic number"), and of course `null` (which is as anonymous as T/F but has the advantage that it merely fails to convey what reference is null, as opposed to in which of two directions the behavior is changed). Arguably `null` must not be encouraged.

I will try using less deprecatedName for magic names like `getOrElse`, `orElse` which take a single boolean param. Instead, target just the name and signature. I didn't like the asymmetry of sprinkling the annotation on just certain API.

I used to exclude constructors; that idea was to exclude mere wrappers. But it's hard to know what semantically merely wraps, as opposed to `Map(isThreadSafe: Boolean)`.

I will mention deprecatedName wherever possible.

som-snytt commented:
Worth documenting that refchecks compares `SomeClass`: when compiling standard lib, `SomeModule` is different from the synthetic companion; also, the `new Some` rewrite happens the symbol of the `Apply` is stale. Flying only by instrumentation, comparing the companion seems to work.

som-snytt commented:
Also unsatisfactory is variously named fold arg:
```
linked.fold(none = false)(accessWithin)
c.alias.fold(ifEmpty = false)(a => a == parsedLineWord)
clazz.methods.asScala.foldLeft(z=false)
```
Also Intelli already lints with action:
![image](https://github.com/scala/scala/assets/369425/5fa1aa53-292b-4e39-8958-563e83baa73a)

som-snytt committed:
Use deprecatedName for some boolean parameters

som-snytt committed:
Use named literal boolean args in project

som-snytt commented:
`assert` could be specialized.

lrytz commented on src/compiler/scala/tools/nsc/transform/async/TransformUtils.scala:
```diff
@@ -89,7 +89,7 @@ private[async] trait TransformUtils extends AsyncTransformStates {
   def literalUnit: Tree = Literal(Constant(())).setType(definitions.UnitTpe) // a def to avoid sharing trees
   def literalBoxedUnit: Tree = gen.mkAttributedRef(definitions.BoxedUnit_UNIT)
-  def literalBool(b: Boolean): Tree = Literal(Constant(b)).setType(definitions.BooleanTpe)
+  def literalBool[B <: Boolean](b: B): Tree = Literal(Constant(b)).setType(definitions.BooleanTpe)
```
Is there a reason not to use `@deprecatedName` here?

som-snytt replied:
this was a case where this works just as well (since it is boxed for Constant, assuming it is not reboxed), so this was the experiment.

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1697,6 +1697,8 @@ abstract class RefChecks extends Transform {
       toConstructor(tree.pos, tree.tpe)
     }
 
+    private lazy val ArrayApplyBoolean = ArrayModule_apply(BooleanTpe)
```
Can we use `@deprecatedName` instead?

som-snytt replied:
I liked the idea of putting the burden on the lint code, unless the annotation made compelling sense.

som-snytt replied:
Array.apply gets the lrytz exclusion. `Boolean*` is not a `Boolean`.

lrytz commented on test/files/neg/named-booleans.scala:
```diff
@@ -0,0 +1,47 @@
+  val y = Some(false)
+  val z = Option(false)
+  val w = (true, false)
+  val v = c g true  // nowarn?
```
This probably takes a different code path in the parser, see `makeBinop`

som-snytt replied:
that's right. I decided it's ok because whatever. I did see the idiom in the codebase!

lrytz reviewed (commented):
It occured to me while looking at this, making an exception for the first parameter might be a useful heuristic. It would cover things like `assert` or `literalBool`. But it might also be too surprising - what do you think?

lrytz commented:
Re-arranged the commits (with a small diff: https://github.com/scala/scala/compare/77320f3bffca6d6e295a137278e481f8d405e371..e7da16c208b25b4919fca0de562a5da730318d96)

> making an exception for the first parameter might be a useful heuristic

That could be narrowed down to "if the first parameter is a boolean and there are no other boolean parameters".

som-snytt commented:
OK I'll just book the education for getting the symbol of an array apply. You may issue me a paper certificate.

We can squash, I was just busy yesterday and tentative. I'm glad for your simplifications.

lrytz commented:
👍 

What do you think about the heuristic for `assert` or `boolConst` to skip the warning if there's a single boolean parameter on the first position?

som-snytt commented:
Maybe it makes sense in a mysterious way. (that I don't understand)

Trying to think of a counter-example, `Option.when` would not need deprecatedName.

Is the explanation, "A single leading boolean parameter must naturally follow the name of the method, and therefore needs no name."

A perverse API would be `f(flag: Boolean, data: String)` where the flag inverts the behavior. (Then the method name does not explain what the boolean is for.)

som-snytt commented on src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala:
```diff
@@ -29,7 +29,7 @@ trait MemberHandlers {
 
   private def codegenln(leadingPlus: Boolean, xs: String*): String = codegen(leadingPlus, (xs ++ Array("\n")): _*)
-  private def codegenln(xs: String*): String = codegenln(true, xs: _*)
+  private def codegenln(xs: String*): String = codegenln(leadingPlus = true, xs: _*)
```
example, albeit minor, of leading bool. In this case, it is tempting because the other param is repeating; also, the flag is for `leadingPlus`, so it looks right to come first.

som-snytt commented:
Sample of things with leading boolean parameter. Some are the `assert` paradigm.

Aside from "how it reads," such as `flagIf(true, 42)`, do you ever expect to be called with a literal?

`new ResetAttrs(brutally = true, ignore)`

```
src/repl/scala/tools/nsc/interpreter/Interface.scala:  def completionCandidates(filter: Boolean, tabCount: Int): (Int, List[CompletionCandidate])

src/scaladoc/scala/tools/nsc/doc/html/page/Entity.scala:    def inside(hasLinks: Boolean, nameLink: String = ""): Elems =

src/reflect/scala/reflect/internal/tpe/TypeConstraints.scala:    @inline def toBound(hi: Boolean, tparam: Symbol) = if (hi) tparam.info.upperBound else tparam.info.lowerBound

src/reflect/scala/reflect/internal/Symbols.scala:  case class SymbolOps(isFlagRelated: Boolean, mask: Long)

src/reflect/scala/reflect/internal/Kinds.scala:      def kindCheck(cond: Boolean, f: KindErrors => KindErrors): Unit =

src/reflect/scala/reflect/io/File.scala:  def writer(append: Boolean, codec: Codec): OutputStreamWriter =

src/repl-frontend/scala/tools/nsc/interpreter/shell/ILoop.scala:  def asyncEcho(async: Boolean, msg: => String): Unit = {

src/repl-frontend/scala/tools/nsc/interpreter/shell/LoopCommands.scala:  case class Result(keepRunning: Boolean, lineToRecord: Option[String])

src/scalap/scala/tools/scalap/JavaWriter.scala:  def flagsToStr(clazz: Boolean, flags: Int): String = {

src/scalap/scala/tools/scalap/Classfile.scala:  case class Member(field: Boolean, flags: Int, name: Int, tpe: Int, attribs: List[Attribute])

src/compiler/scala/tools/tasty/TastyHeaderUnpickler.scala:  private def check(cond: Boolean, msg: => String): Unit = {

src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala:  final protected def unsupportedWhen(cond: Boolean, msg: => String)(implicit ctx: Context): Unit =

src/compiler/scala/tools/nsc/tasty/TreeUnpickler.scala:    private def normalizeName(isType: Boolean, name: TastyName)(implicit ctx: Context): TastyName = {

src/compiler/scala/tools/nsc/tasty/bridge/ContextOps.scala:  @inline final def assert(assertion: Boolean, msg: => Any): Unit =

src/compiler/scala/tools/nsc/ast/TreeGen.scala:  def mkManifestFactoryCall(full: Boolean, constructor: String, tparg: Type, args: List[Tree]): Tree =

src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:    def makeBinop(isExpr: Boolean, left: Tree, op: TermName, right: Tree, opPos: Position, targs: List[Tree] = Nil): Tree = {
src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:    def finishBinaryOp(isExpr: Boolean, opinfo: OpInfo, rhs: Tree): Tree = {
src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:    def reduceStack(isExpr: Boolean, base: List[OpInfo], top: Tree): Tree = {
src/compiler/scala/tools/nsc/ast/parser/Parsers.scala:    def reduceStack(isExpr: Boolean, base: List[OpInfo], top: Tree, opPrecedence: Precedence, leftAssoc: Boolean): Tree = {

src/compiler/scala/tools/nsc/ast/Trees.scala:  private class ResetAttrs(brutally: Boolean, leaveAlone: Tree => Boolean)

src/compiler/scala/tools/nsc/backend/jvm/BTypes.scala:  final case class InlineInfo(isEffectivelyFinal: Boolean,

src/compiler/scala/tools/nsc/backend/jvm/BCodeBodyBuilder.scala:    def flagIf(b: Boolean, flag: Int): Int = if (b) flag else 0

src/compiler/scala/tools/nsc/backend/jvm/analysis/ProdConsAnalyzer.scala:  override def newParameterValue(isInstanceMethod: Boolean, local: Int, tp: Type): SourceValue = {

src/compiler/scala/tools/nsc/Global.scala:      def checkPhaseSettings(including: Boolean, specs: Seq[String]*) = {

src/compiler/scala/tools/nsc/typechecker/splain/SplainData.scala:    def finishSearch(success: Boolean, expectedType: Type): Unit = {

src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:      private def setErrorOnLastTry(lastTry: Boolean, tree: Tree) = if (lastTry) setError(tree)

src/compiler/scala/tools/nsc/typechecker/TreeCheckers.scala:  def assertFn(cond: Boolean, msg: => Any) =

src/compiler/scala/tools/nsc/typechecker/StdAttachments.scala:  case class MacroRuntimeAttachment(delayed: Boolean, typerContext: Context, macroContext: Option[MacroContext])

src/compiler/scala/tools/nsc/typechecker/Implicits.scala:    private def checkCompatibility(fast: Boolean, tp0: Type, pt0: Type): Boolean = {

src/library/scala/collection/immutable/RedBlackTree.scala:  private[this] def mkTree[A, B](isBlack: Boolean, key: A, value: B, left: Tree[A, B], right: Tree[A, B]) = {

src/library/scala/sys/process/ProcessImpl.scala:  private[process] abstract class PipeThread(isSink: Boolean, labelFn: () => String) extends Thread {
src/library/scala/sys/process/BasicIO.scala:    def apply[T](nonzeroException: Boolean, capacity: Integer): LazilyListed[T] = {


```

som-snytt committed:
Improvements

som-snytt commented:
OK, I corrected the code in refchecks and implemented the exclusion.

IIRC, back in September, I went back and forth on constructors:
```
new NeedsSigCollector(true)
```
here, the arg is entirely opaque. Obviously, the param flips the behavior of the class somehow.

som-snytt commented:
`readMember(field = true)` obviously needs the name. I did not roll back all the previous edits.

refchecks must decide which param list it's looking at. More convenient is `fn.tpe match { case MethodType(params, _) }` but that breaks boolean parameter detection.

som-snytt commented:
I think the justification is that with one boolean, you can't accidentally swap it. I think you're right that we can enforce some safety, avoid excess noise, and let intellij prompt in other cases.

To your point, or second point, it's up to the person designing the API to use good names for readability, and/or avoid boolean params. `readField` and `readMemberNotField`, whatever.

I remember the complaint that led to some odd constructs in Parsers was passing booleans down the stack. We have better mechanisms now. I wonder what an audit of dotty code now would show.

som-snytt commented on src/library/scala/Array.scala:
```diff
@@ -204,7 +204,7 @@ object Array {
   /** Creates an array of `Boolean` objects */
   // Subject to a compiler optimization in Cleanup, see above.
-  def apply(x: Boolean, xs: Boolean*): Array[Boolean] = {
+  def apply(@deprecatedName x: Boolean, xs: Boolean*): Array[Boolean] = {
```
can be reverted, if it correctly detects repeated boolean is not boolean

lrytz commented on src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:
```diff
@@ -1713,11 +1713,45 @@ abstract class RefChecks extends Transform {
+          }
+        loop(fn, 0)
+      }
+      def isAssertParadigm(params: List[Symbol]): Boolean = !sym.isConstructor && !sym.isCaseApplyOrUnapply && {
```
I'm not convinced we should make a difference between constructors and other methods here.

som-snytt replied:
classes are less likely to have a name that makes its parameter meaningful, unlike a method which is of narrower semantic scope.
```
val EmptyInlineInfo = InlineInfo(false, None, SortedMap.empty, None)
```
I wound up adding names to all four args.

lrytz reviewed (commented):
LGTM otherwise!

som-snytt commented:
I thought I had given it one last scan. Now the diff is long and dreary enough that I tune out rather quickly.

This effort is less about a lint than about how to write a method. Whose idea was it to have more than one parameter anyway?

If you accept my theory mitigating the lrytz exclusion for constructors and case apply, you may enable merging with a green review.

SethTisue commented:
@som-snytt mind updating the PR description to lead with user-facing documentation?

som-snytt commented:
If Keats were alive today (for some interval), he would pen `Ode on the Edit Button`. It would be about impermanence and the fragility of existence, but also affirm our place in the flow of time and change.

I should say, he would pun `Ode on the Edit Button`, because the occasion of the poem would be that his mouse cursor is hovering `on` the edit button, and he is considering whether to click it and therefore enter our experience of temporality and the passing of all things.

Then critics for hundreds of years following would debate whether edit is a button or a menu item, and whether Keats intended it ironically.

retronym commented:
I'm seeing this when building 2.13 from sbt without bootstrapping.

```
[warn] /Users/jz/code/scala/test/junit/scala/reflect/internal/AsSeenFromTest.scala:20:22: Unused import
[warn]   import definitions._
[warn]                      ^
[error] /Users/jz/code/scala/test/junit/scala/tools/nsc/backend/jvm/opt/CallGraphTest.scala:40:37: deprecated parameter name <none> has to be distinct from any other parameter name (deprecated or not).
[error]                     @deprecatedName safeToInline: Boolean, @deprecatedName atInline: Boolean, @deprecatedName atNoInline: Boolean, argInfos: IntMap[ArgInfo] = IntMap.empty): Unit = {
[error]                                     ^
[error] /Users/jz/code/scala/test/junit/scala/tools/nsc/backend/jvm/opt/CallGraphTest.scala:40:76: deprecated parameter name <none> has to be distinct from any other parameter name (deprecated or not).
[error]                     @deprecatedName safeToInline: Boolean, @deprecatedName atInline: Boolean, @deprecatedName atNoInline: Boolean, argInfos: IntMap[ArgInfo] = IntMap.empty): Unit = {
[error]                                                                            ^
[error] /Users/jz/code/scala/test/junit/scala/tools/nsc/backend/jvm/opt/CallGraphTest.scala:40:111: deprecated parameter name <none> has to be distinct from any other parameter name (deprecated or not).
[error]                     @deprecatedName safeToInline: Boolean, @deprecatedName atInline: Boolean, @deprecatedName atNoInline: Boolean, argInfos: IntMap[ArgInfo] = IntMap.empty): Unit = {
[error]                                                                                                               ^
[error] /Users/jz/code/scala/test/junit/scala/tools/nsc/backend/jvm/opt/InlineSourceMatcherTest.scala:32:47: deprecated parameter name <none> has to be distinct from any other parameter name (deprecated or not).
[error]   case class E(regex: String, @deprecatedName negated: Boolean = false, @deprecatedName terminal: Boolean = true)
[error]                                               ^
[error] /Users/jz/code/scala/test/junit/scala/tools/nsc/backend/jvm/opt/InlineSourceMatcherTest.scala:32:89: deprecated parameter name <none> has to be distinct from any other parameter name (deprecated or not).
[error]   case class E(regex: String, @deprecatedName negated: Boolean = false, @deprecatedName terminal: Boolean = true)
[error]                                                                                         ^
[warn] 10 warnings found
[error] 5 errors found
[error] (junit / Test / compileIncremental) Compilation failed
[error] Total time: 42 s, completed 15 Dec 2023, 2:06:53 pm
```

som-snytt commented:
@retronym I thought I fixed that, but maybe I broke it instead.

Probably I should have silenced warnings the old-fashioned way until re-starr; I wound up not making accommodations in regular code (improving the heuristic instead), but left some in tests.

I'll review this PR (now) for those changes that should wait for re-starr. It's been so long since I had to wait for re-starr, I kind of forgot what it was like, the thrill and the sense of anticipation. Oh wait, that is Christmas morning.

som-snytt commented on src/library/scala/concurrent/SyncChannel.scala:
```diff
@@ -19,12 +19,14 @@ package scala.concurrent
-  private[this] var pendingWrites = List[(A, SyncVar[Boolean])]()
+  private final val Signal = ()
+  private type Signal = Unit
+  private[this] var pendingWrites = List[(A, SyncVar[Signal])]()
```
I thought I had reverted this change. It may not matter, but there's no reason to update deprecated classes. It's a candidate for reversion.

SethTisue commented:
I had a good experience with this in a closed-source codebase. I got two helpful warnings and no annoying ones.

som-snytt commented:
I need it to also lint
```
  x.withPartial(false)
                ^
Did you mean true?
```

som-snytt commented on src/compiler/scala/tools/nsc/transform/patmat/MatchTreeMaking.scala:
```diff
@@ -24,10 +24,10 @@ import scala.tools.nsc.Reporting.WarningCategory
   import global._, definitions._, CODE._
 
-  final case class Suppression(suppressExhaustive: Boolean, suppressUnreachable: Boolean)
+  final case class Suppression private (suppressExhaustive: Boolean, suppressUnreachable: Boolean)
```
This is charming but makes `copy` hard to use (under `-Xsource:3`).

SethTisue commented:
It's likely we'll adjust this for 2.13.14; let's discuss on #10704
